mlflow/openai/_openai_autolog.py

```diff
@@ -0,0 +1,224 @@
+        if param := kwargs.get(param_name):
+            return param
+        input_example_exc = MlflowException(
+            "Inference function signature changes, please contact MLflow team to "
+            "fix OpenAI autologging.",
+        )
+    else:
+        input_example_exc = MlflowException(
+            f"Unsupported OpenAI task. Only support {list(model_class_param_name_mapping.keys())}."
+        )
+    _logger.warning(
+        f"Failed to gather input example of model {model.__class__.__name__} "
+        f"due to error: {input_example_exc}."
+    )
+
+
+def _convert_data_to_dict(data, key):
+    if isinstance(data, dict):
+        return {f"{key}-{k}": v for k, v in data.items()}
+    if isinstance(data, list):
+        return {key: data}
+    if isinstance(data, str):
+        return {key: [data]}
+    raise MlflowException("Unsupported data type.")
+
+
+def _combine_input_and_output(input, output):
+    """
+    Combine input and output into a single dictionary
+    """
+    result = {}
+    if input:
+        result.update(_convert_data_to_dict(input, "input"))
+    if output:
+        output = [output.model_dump(mode="json")]
+        result.update(_convert_data_to_dict(output, "output"))
+    return result
+
+
+@contextmanager
+def _set_api_key_env_var(client):
+    """
+    Gets the API key from the client and temporarily set it as an environment variable
+    """
+    api_key = client.api_key
+    original = os.environ.pop("OPENAI_API_KEY", None)
+    os.environ["OPENAI_API_KEY"] = api_key
+    yield
+    if original is not None:
+        os.environ["OPENAI_API_KEY"] = original
+    else:
+        os.environ.pop("OPENAI_API_KEY")
+
+
+def patched_call(original, self, *args, **kwargs):
+    from openai import Stream
+
+    run_id = getattr(self, "run_id", None)
+    active_run = mlflow.active_run()
+    mlflow_client = mlflow.MlflowClient()
+    if run_id is None:
+        # only log the tags once
+        extra_tags = get_autologging_config(mlflow.openai.FLAVOR_NAME, "extra_tags", None)
+        # include run context tags
+        resolved_tags = context_registry.resolve_tags(extra_tags)
+        tags = _resolve_extra_tags(mlflow.openai.FLAVOR_NAME, resolved_tags)
+        if active_run:
+            run_id = active_run.info.run_id
+            mlflow_client.log_batch(
+                run_id=run_id,
+                tags=[RunTag(key, str(value)) for key, value in tags.items()],
+            )
+        else:
+            run = mlflow_client.create_run(
+                experiment_id=_get_experiment_id(),
+                tags=tags,
+            )
+            run_id = run.info.run_id
+
+    with disable_autologging():
+        result = original(self, *args, **kwargs)
+
+    class _OpenAIJsonEncoder(json.JSONEncoder):
+        def default(self, o):
+            from openai._types import NotGiven
+
+            if isinstance(o, NotGiven):
+                return str(o)
+
+            return super().default(o)
+
```

Comment:
Can this class be pulled out from this function?

Reply:
Can we do:

```
try:
    return super().default(o)
except TypeError:
    return str(o)
```

?

Reply:
i think that works too